packages/transaction-controller/CHANGELOG.md

```diff
@@ -7,6 +7,32 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- `estimateGasBuffered()` now accepts optional networkClientId as its last param which specifies the network client that should be used to estimate the required gas plus buffer for the given transaction ([#3643](https://github.com/MetaMask/core/pull/3643))
+- `getNonceLock()` now accepts optional networkClientId as its last param which specifies which the network client's nonceTracker should be used to determine the next nonce. ([#3643](https://github.com/MetaMask/core/pull/3643))
+  - When used with the `isMultichainEnabled` feature flag on and with networkClientId specified, this method will also restrict acquiring the next nonce by chainId, i.e. if this method is called with two different networkClientIds on the same chainId, only the first call will return immediately with a lock from its respective nonceTracker with the second call being blocked until the first caller releases its lock
+- `EtherscanRemoteTransactionSource` now enforces a 5 second delay between requests to avoid rate limiting ([#3643](https://github.com/MetaMask/core/pull/3643))
+- `TransactionMeta` type now specifies an optional `networkClientId` field ([#3643](https://github.com/MetaMask/core/pull/3643))
+- `startIncomingTransactionPolling()` now accepts an optional array of `networkClientIds`. When provided, the `IncomingTransactionHelpers` for those networkClientIds will be started. If empty or not provided, the global `IncomingTransactionHelper` will be started. ([#3643](https://github.com/MetaMask/core/pull/3643))
+- `stopIncomingTransactionPolling()` now accepts an optional array of `networkClientIds`. When provided, the `IncomingTransactionHelpers` for those networkClientIds will be stoppped. If empty or not provided, the global `IncomingTransactionHelper` will be stopped. ([#3643](https://github.com/MetaMask/core/pull/3643))
+
+### Added
+
+- **BREAKING:** Constructor now expects a `getNetworkClientRegistry` callback function ([#3643](https://github.com/MetaMask/core/pull/3643))
+- **BREAKING:** Messenger now requires `NetworkController:stateChange` to be an allowed event ([#3643](https://github.com/MetaMask/core/pull/3643))
+- **BREAKING:** Messenger now requires `NetworkController:findNetworkClientByChainId` and `NetworkController:getNetworkClientById` actions ([#3643](https://github.com/MetaMask/core/pull/3643))
+- Adds a feature flag parameter `isMultichainEnabled` passed via the constructor (and defaulted to false), which when passed a truthy value will initialize `IncomingTransactionHelper`, `PendingTransactionTracker`, `NonceTracker` per networkClientId from the NetworkController’s `networkClientRegistry` and a EtherScanRemoteTransactionSource helper per chainId in the registry. ([#3643](https://github.com/MetaMask/core/pull/3643))
```

Comment:
As above, perhaps there is a way we can explain the behavior more generally? I think this basically makes it possible to track transactions for multiple networks? Also, isn't it true that for some methods (I don't remember which ones), `networkClientId` will only take effect if `isMultichainEnabled` is passed?

Reply:
behavior more generally explained with less private implementation detail and method behavior when feature flag is on/off better documented

https://github.com/MetaMask/core/pull/3925/commits/b333b53be39f803b160042699fa270894339442e

Reply:
Nice!